File: src/main/java/coffee/amo/interactionkey/events/CommonEvents.java
package coffee.amo.interactionkey.events;

import coffee.amo.interactionkey.InteractionKey;
import coffee.amo.interactionkey.api.event.InteractionKeyPressedEvent;
import coffee.amo.interactionkey.client.ClientUtil;
import coffee.amo.interactionkey.compat.ItemPhysicCompat;
import coffee.amo.interactionkey.config.InteractionKeyConfig;
import coffee.amo.interactionkey.net.PacketHandler;
import net.minecraft.client.Minecraft;
import net.minecraft.core.BlockPos;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.entity.item.ItemEntity;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.context.UseOnContext;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.EntityHitResult;
import net.minecraft.world.phys.HitResult;
import net.minecraftforge.common.ForgeMod;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.ModList;
import net.minecraftforge.fml.common.Mod;
import team.creative.itemphysic.ItemPhysic;
import team.creative.itemphysic.common.packet.PickupPacket;

import java.util.ArrayList;
import java.util.List;

@Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.FORGE, modid = InteractionKey.MODID)
public class CommonEvents {

    @SubscribeEvent
    public static void onInteractionKeyPressed(InteractionKeyPressedEvent event) {
        if(event.getEntity().getItemInHand(InteractionHand.MAIN_HAND).getItem() instanceof BlockItem) return;
        HitResult hitResult = event.getEntity().pick(event.getEntity().getAttributeValue(ForgeMod.REACH_DISTANCE.get()), 0, false);
        if(hitResult.getType() == HitResult.Type.ENTITY){
            if(((EntityHitResult)hitResult).getEntity() instanceof ItemEntity){
                if(event.getEntity().level.isClientSide && ModList.get().isLoaded("itemphysic")){
                    ItemPhysicCompat.doPickup((ItemEntity) ((EntityHitResult)hitResult).getEntity(), event.getEntity());
                    return;
                }
            }
            event.getEntity().interactOn(event.getEntity(), InteractionHand.MAIN_HAND);
            return;
        }
        if(InteractionKeyConfig.enableVanillaInteract.get()){
            // get the block the player is looking at on the server
            if(hitResult.getType() == HitResult.Type.BLOCK){
                BlockHitResult blockHit = (BlockHitResult) hitResult;
                // if the player is looking at a block, interact with it
                if(event.getEntity().level.isClientSide) ClientUtil.useItemHack();
                return;
           }
        }
        if(event.getEntity().level.isClientSide){
            if(ModList.get().isLoaded("itemphysic")){
                ItemPhysicCompat.playerPickupInteract(event.getEntity());
                return;
            }
        }
    }
}

File: src/main/kotlin/coffee/amo/interactionkey/events/CommonEvents.kt
package coffee.amo.interactionkey.events

import coffee.amo.interactionkey.InteractionKey
import coffee.amo.interactionkey.api.InteractableEntityHolder
import coffee.amo.interactionkey.api.event.InteractionKeyPressedEvent
import coffee.amo.interactionkey.client.ClientUtil
import coffee.amo.interactionkey.compat.ItemPhysicCompat
import coffee.amo.interactionkey.config.InteractionKeyConfig
import net.minecraft.core.Direction
import net.minecraft.network.chat.Component
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.item.ItemEntity
import net.minecraft.world.item.BlockItem
import net.minecraft.world.phys.BlockHitResult
import net.minecraft.world.phys.EntityHitResult
import net.minecraft.world.phys.HitResult
import net.minecraft.world.phys.Vec3
import net.minecraftforge.common.ForgeMod
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.eventbus.api.SubscribeEvent
import net.minecraftforge.fml.ModList
import net.minecraftforge.fml.common.Mod.EventBusSubscriber

@EventBusSubscriber(bus = EventBusSubscriber.Bus.FORGE, modid = InteractionKey.MODID)
object CommonEvents {
    @SubscribeEvent
    fun onInteractionKeyPressed(event: InteractionKeyPressedEvent) {
        if(event.entityID != InteractionKey.CHECK_INT){
            val entity = event.player!!.level.getEntity(event.entityID!!);
            if(entity is ItemEntity){
                InteractableEntityHolder(entity, null, null, 0).interact(event.player)
                return;
            }
            try {
                InteractableEntityHolder(entity!!, null, null, 0).interact(event.player)
                return;
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        if(event.blockPos != InteractionKey.CHECK_POS){
            if(event.player!!.level.isClientSide) return
            event.player!!.level.getBlockState(event.blockPos!!).use(event.player.level, event.player, InteractionHand.MAIN_HAND, BlockHitResult(
                Vec3.atCenterOf(event.blockPos), Direction.UP, event.blockPos, false))
            return;
        }
        //        if(event.getEntity().level.isClientSide){
//            if(ModList.get().isLoaded("itemphysic")){
//                ItemPhysicCompat.playerPickupInteract(event.getEntity());
//                return;
//            }
//        }
    }

    fun init(){
        MinecraftForge.EVENT_BUS.register(this)
    }
}